backend/logic/deploy/bucket/gcp.js

```diff
@@ -0,0 +1,139 @@
+const path = require('path')
+const trimStart = require('lodash/trimStart')
+const { Storage } = require('@google-cloud/storage')
+
+const { BUCKET_PREFIX } = require('../../../utils/const')
+const { walkDir } = require('../../../utils/filesystem')
+const { assert } = require('../../../utils/validators')
+const { getLogger } = require('../../../utils/logger')
+
+const log = getLogger('logic.deploy.bucket.gcp')
+
+let cachedClient = null
+
+/**
+ * Check if GCP is configured and we can deploy to it
+ *
+ * @param args {Object}
+ * @param args.networkConfig {Object} - Decrypted networkConfig object
+ * @returns {bool} - if we can deploy
+ */
+function isAvailable({ networkConfig }) {
+  return !!networkConfig.gcpCredentials
+}
+
+/**
+ * Configure this singleton for a use
+ *
+ * @param args {Object}
+ * @param args.networkConfig {Object} - Decrypted networkConfig object
+ * @param args.credentials {Object} - GCP credentials (if networkConfig not provided)
+ */
+function configure({ networkConfig, credentials }) {
+  let creds = credentials ? credentials : networkConfig.gcpCredentials
+
+  if (typeof creds === 'string') {
+    creds = JSON.parse(creds)
+  }
+
+  cachedClient = new Storage({
+    projectId: creds.project_id, // Shrug
+    credentials: creds
+  })
+}
+
+/**
+ * Deploy OutputDir to a properly named bucket
+ *
+ * @param args {Object}
+ * @param args.networkConfig {Object} - Decrypted networkConfig object
+ * @param args.OutputDir {string} - The directory containing the shop build
+ */
+async function deploy({ shop, networkConfig, OutputDir }) {
+  assert(!!networkConfig, 'networkConfig must be provided')
+  assert(!!OutputDir, 'OutputDir must be provided')
+
+  const bucketName = `${BUCKET_PREFIX}${shop.authToken}`
```

Comment:
should we include the networkId to avoid any possible collision between shop deployments on different network but having the same authToken?

Reply:
I hadn't considered that.  Do we allow any `authToken` collisions in an instance?  I don't think so.  But since we may have multiple instances(for us, staging/prod), `BUCKET_PREFIX` should probably at least reflect that.

Reply:
I added `NODE_ENV` into the prefixes.

Reply:
We don't allow collision on an instance (we check the authToken does not already exist in the DB before using it for a new shop).

I believe NODE_ENV is set to "production" for both staging and prod. I'd be slightly inclined to use networkId.

Reply:
> I believe NODE_ENV is set to "production" for both staging and prod so I'd be inclined to use networkId.

Well, that makes this tricky.  I still think there's a good possibility in the future for an individual node to support multiple networks, so I'm hesitant to use networkId.  Though... we're depending on networkConfig for credentials right now.  Maybe that is a good choice.